chamge/amqp-arbiter-backend.c

```diff
@@ -32,65 +33,294 @@ G_DEFINE_TYPE (ChamgeAmqpArbiterBackend, chamge_amqp_arbiter_backend, CHAMGE_TYP
+
+  /* if the content-type isn't 'application/json', let's drop */
+  if (!(envelope.message.properties._flags & AMQP_BASIC_CONTENT_TYPE_FLAG)
+      || strlen ("application/json") !=
+      envelope.message.properties.content_type.len
+      || g_ascii_strncasecmp ("application/json",
+          envelope.message.properties.content_type.bytes,
+          envelope.message.properties.content_type.len)
+      ) {
+    g_debug ("invalid content type %s",
+        (gchar *) envelope.message.properties.content_type.bytes);
+
+    goto out;
+  }
+
+  if (envelope.message.properties.reply_to.len > 0 &&
+      (strlen (envelope.message.properties.reply_to.bytes) >=
+          envelope.message.properties.reply_to.len)) {
+    *reply_to = g_strndup (envelope.message.properties.reply_to.bytes,
+        envelope.message.properties.reply_to.len);
+    g_debug ("replay to == %s", *reply_to);
+  } else {
+    g_debug ("there is no replay to in property");
+  }
+
+  if (envelope.message.properties.correlation_id.len > 0 &&
+      (strlen (envelope.message.properties.correlation_id.bytes) >=
+          envelope.message.properties.correlation_id.len)) {
+    gchar *id = g_strndup (envelope.message.properties.correlation_id.bytes,
+        envelope.message.properties.correlation_id.len);
+    *correlation_id = atoi (id);
+    g_debug ("correlation id == %d", *correlation_id);
+  } else {
+    g_debug ("there is no correlation id to in property");
+  }
+
+  g_debug ("Content-type: %.*s",
+      (int) envelope.message.properties.content_type.len,
+      (char *) envelope.message.properties.content_type.bytes);
+
+  if (envelope.message.body.len > 0 && envelope.message.body.bytes != NULL) {
+    *body = g_strndup (envelope.message.body.bytes, envelope.message.body.len);
+  }
 
-  g_debug ("parsed amqp uri (host: %s, vhost: %s)", amqp_c_info.host,
-      amqp_c_info.vhost);
+  amqp_destroy_envelope (&envelope);
+  return CHAMGE_RETURN_OK;
+
+out:
+  amqp_destroy_envelope (&envelope);
+  return CHAMGE_RETURN_FAIL;
+}
 
-  amqp_socket_open (self->amqp_socket, amqp_c_info.host, amqp_c_info.port);
+static ChamgeReturn
+_amqp_rpc_respond (amqp_connection_state_t amqp_conn, const gchar * response,
+    const gchar * reply_to, guint correlation_id)
+{
+  amqp_basic_properties_t amqp_props;
 
+  if (_amqp_property_set (&amqp_props, correlation_id,
+          NULL) != CHAMGE_RETURN_OK) {
+    return CHAMGE_RETURN_FAIL;
+  }
   /*
-   * TODO: The authentication method should be EXTERNAL
-   * if we don't want to share user/passwd
+   * publish
    */
-  amqp_login (self->amqp_conn, amqp_c_info.vhost, 0, 131072, 0,
-      AMQP_SASL_METHOD_PLAIN, amqp_c_info.user, amqp_c_info.password);
-  amqp_channel_open (self->amqp_conn, amqp_channel);
-  amqp_get_rpc_reply (self->amqp_conn);
+  g_debug ("publishing to [%s]", reply_to);
+  g_debug ("      correlation id : %d body : \"%s\"", correlation_id, response);
+  amqp_basic_publish (amqp_conn, 1, amqp_cstring_bytes (""),
```

Comment:
If we need to have common utility function, it should be able to accept *channel*.